gateway/websocket_request_context.go

```diff
@@ -146,99 +142,62 @@ func (wrc *websocketRequestContext) buildProtocolContextFromHTTPRequest(
 	}
 
 	// Build protocol context for the selected endpoint
-	protocolCtx, protocolCtxSetupErrObs, err := wrc.protocol.BuildRequestContextForEndpoint(wrc.context, wrc.serviceID, selectedEndpoint, httpReq)
+	protocolCtx, protocolCtxSetupErrObs, err := wrc.protocol.BuildWebsocketRequestContextForEndpoint(
+		wrc.context,
+		wrc.serviceID,
+		selectedEndpoint,
+		httpReq,
+	)
 	if err != nil {
 		// error encountered: use the supplied observations as protocol observations.
 		wrc.updateProtocolObservations(&protocolCtxSetupErrObs)
 		logger.Error().Err(err).Str("endpoint_addr", string(selectedEndpoint)).Msg("Failed to build protocol context for websocket endpoint")
 		return fmt.Errorf("failed to build protocol context for websocket endpoint: %w", err)
 	}
+	wrc.logger = wrc.logger.With("endpoint_addr", selectedEndpoint)
 
 	wrc.protocolCtx = protocolCtx
 	logger.Info().Msgf("Successfully built protocol context for websocket endpoint: %s", selectedEndpoint)
-
-	// If no error occurred, update protocol connection observations
-	// This will be used to broadcast connection observations at the end of the connection lifecycle.
-	wrc.updateProtocolObservations(&endpointLookupObs)
-
 	return nil
 }
 
 // handleWebsocketRequest establishes the websocket connection and starts the bridge,
 // which handles the message processing loop and sends message observations to the gateway.
+//
+// It also starts a goroutine to listen for message processing notifications from the bridge.
+//
 // This method blocks until the WebSocket connection terminates.
 func (wrc *websocketRequestContext) handleWebsocketRequest(
 	httpRequest *http.Request,
 	httpResponseWriter http.ResponseWriter,
 ) error {
 	logger := wrc.logger.With("method", "handleWebsocketRequest")
 
-	// Create the websocket bridge and start it.
-	completionChan, err := wrc.startWebSocketBridge(httpRequest, httpResponseWriter)
-	if err != nil {
-		logger.Error().Err(err).Msg("❌ Failed to create websocket bridge.")
-		return err
-	}
-
-	// Wait for the bridge to complete (blocks until WebSocket connection terminates)
-	// in order to allow publishing observations for the connection duration.
-	<-completionChan
+	// Create separate channels for establishment and closure observations
+	establishmentObservationsChan := make(chan *protocolobservations.Observations, 1)
+	closureObservationsChan := make(chan *protocolobservations.Observations, 1)
 
-	logger.Info().Msg("🔌 WebSocket connection terminated, broadcasting final connection observations")
+	// Start listening for connection observations in separate goroutines
+	go wrc.listenForEstablishmentObservations(establishmentObservationsChan)
```

Comment:
There is some protocol-internal logic leaking here: all of these are protocol observations, produced by protocol, over a single channel that contains a single protocol observation type (e.g. WebSocketObservation)